File: cpp/tests/test_integration.cpp
#include <catch2/catch_all.hpp>
#include <iostream>
#include <string>
#include <vector>
#include <memory>

#include "magda_cpp/pipeline.hpp"
#include "magda_cpp/models.h"

/**
 * @brief Integration tests for MAGDA C++ implementation with real API calls
 */
TEST_CASE("Basic track operations integration", "[integration][track]") {
    // Get API key from environment
    const char* api_key = std::getenv("OPENAI_API_KEY");
    if (!api_key) {
        SKIP("OPENAI_API_KEY not set - skipping integration tests");
    }

    magda::MAGDAPipeline pipeline(api_key);

    std::vector<std::string> track_prompts = {
        "Create a new track called 'Electric Guitar'",
        "Add a new audio track named 'Acoustic Bass'",
        "Delete the track 'Old Drums'",
        "Crea una nueva pista llamada 'Piano'",
        "Ajoute une piste MIDI nommée 'Synth'"
    };

    for (const auto& prompt : track_prompts) {
        std::cout << "\nTesting track operation: " << prompt << std::endl;

        auto result = pipeline.processPrompt(prompt);
        REQUIRE(result.has_value());
        REQUIRE(result->operations.size() > 0);

        // Check for track operations
        bool has_track_op = false;
        for (const auto& op : result->operations) {
            if (op.operation_type == magda::OperationType::CREATE_TRACK) {
                has_track_op = true;
                break;
            }
        }
        REQUIRE(has_track_op);

        std::cout << "✓ Track operation passed" << std::endl;
    }
}

TEST_CASE("Volume operations integration", "[integration][volume]") {
    const char* api_key = std::getenv("OPENAI_API_KEY");
    if (!api_key) {
        SKIP("OPENAI_API_KEY not set - skipping integration tests");
    }

    magda::MAGDAPipeline pipeline(api_key);

    std::vector<std::string> volume_prompts = {
        "Set the volume of track 'Guitar' to -6dB",
        "Increase the volume of 'Bass' by 3dB",
        "Mute the 'Drums' track",
        "Ajusta el volumen de 'Piano' a -3dB",
        "Réduis le volume de 'Synth' de 2dB"
    };

    for (const auto& prompt : volume_prompts) {
        std::cout << "\nTesting volume operation: " << prompt << std::endl;

        auto result = pipeline.processPrompt(prompt);
        REQUIRE(result.has_value());
        REQUIRE(result->operations.size() > 0);

        // Check for volume operations
        bool has_volume_op = false;
        for (const auto& op : result->operations) {
            if (op.operation_type == magda::OperationType::SET_VOLUME) {
                has_volume_op = true;
                break;
            }
        }
        REQUIRE(has_volume_op);

        std::cout << "✓ Volume operation passed" << std::endl;
    }
}

TEST_CASE("Effect operations integration", "[integration][effect]") {
    const char* api_key = std::getenv("OPENAI_API_KEY");
    if (!api_key) {
        SKIP("OPENAI_API_KEY not set - skipping integration tests");
    }

    magda::MAGDAPipeline pipeline(api_key);

    std::vector<std::string> effect_prompts = {
        "Add reverb to the 'Guitar' track",
        "Remove the delay from 'Bass'",
        "Set the reverb wet level to 50% on 'Guitar'",
        "Füge einen Chorus-Effekt zur 'Piano'-Spur hinzu",
        "Ajoute un filtre passe-bas à la piste 'Synth'"
    };

    for (const auto& prompt : effect_prompts) {
        std::cout << "\nTesting effect operation: " << prompt << std::endl;

        auto result = pipeline.processPrompt(prompt);
        REQUIRE(result.has_value());
        REQUIRE(result->operations.size() > 0);

        // Check for effect operations
        bool has_effect_op = false;
        for (const auto& op : result->operations) {
            if (op.operation_type == magda::OperationType::ADD_EFFECT) {
                has_effect_op = true;
                break;
            }
        }
        REQUIRE(has_effect_op);

        std::cout << "✓ Effect operation passed" << std::endl;
    }
}

TEST_CASE("Clip operations integration", "[integration][clip]") {
    const char* api_key = std::getenv("OPENAI_API_KEY");
    if (!api_key) {
        SKIP("OPENAI_API_KEY not set - skipping integration tests");
    }

    magda::MAGDAPipeline pipeline(api_key);

    std::vector<std::string> clip_prompts = {
        "Create a 4-bar clip on track 'Guitar'",
        "Delete the clip at bar 8 on 'Bass'",
        "Move the clip from bar 4 to bar 12 on 'Drums'",
        "Crea un clip de 8 compases en la pista 'Piano'",
        "Crée un clip de 2 mesures sur la pista 'Synth'"
    };

    for (const auto& prompt : clip_prompts) {
        std::cout << "\nTesting clip operation: " << prompt << std::endl;

        auto result = pipeline.processPrompt(prompt);
        REQUIRE(result.has_value());
        REQUIRE(result->operations.size() > 0);

        // Check for clip operations
        bool has_clip_op = false;
        for (const auto& op : result->operations) {
            if (op.operation_type == magda::OperationType::CREATE_CLIP) {
                has_clip_op = true;
                break;
            }
        }
        REQUIRE(has_clip_op);

        std::cout << "✓ Clip operation passed" << std::endl;
    }
}

TEST_CASE("MIDI operations integration", "[integration][midi]") {
    const char* api_key = std::getenv("OPENAI_API_KEY");
    if (!api_key) {
        SKIP("OPENAI_API_KEY not set - skipping integration tests");
    }

    magda::MAGDAPipeline pipeline(api_key);

    std::vector<std::string> midi_prompts = {
        "Add a C major chord at bar 1 on 'Piano'",
        "Delete the note at beat 2.5 on 'Bass'",
        "Change the velocity of the note at bar 4 to 80",
        "Añade una nota Do en el compás 2 en 'Piano'",
        "Ajoute un accord de La mineur à la mesure 4 sur 'Synth'"
    };

    for (const auto& prompt : midi_prompts) {
        std::cout << "\nTesting MIDI operation: " << prompt << std::endl;

        auto result = pipeline.processPrompt(prompt);
        REQUIRE(result.has_value());
        REQUIRE(result->operations.size() > 0);

        // Check for MIDI operations
        bool has_midi_op = false;
        for (const auto& op : result->operations) {
            if (op.operation_type == magda::OperationType::CREATE_MIDI) {
                has_midi_op = true;
                break;
            }
        }
        REQUIRE(has_midi_op);

        std::cout << "✓ MIDI operation passed" << std::endl;
    }
}

TEST_CASE("Complex multi-operation integration", "[integration][complex]") {
    const char* api_key = std::getenv("OPENAI_API_KEY");
    if (!api_key) {
        SKIP("OPENAI_API_KEY not set - skipping integration tests");
    }

    magda::MAGDAPipeline pipeline(api_key);

    std::vector<std::string> complex_prompts = {
        "Create a new track called 'Lead Guitar', set its volume to -3dB, and add reverb with 40% wet level",
        "Add a new MIDI track 'Piano', create a 4-bar clip, and add a C major chord",
        "Crea una pista llamada 'Bajo', ajusta el volumen a -6dB, y añade un efecto de compresión",
        "Ajoute une piste audio 'Batterie', crée un clip de 8 mesures, et ajoute un filtre passe-haut"
    };

    for (const auto& prompt : complex_prompts) {
        std::cout << "\nTesting complex operation: " << prompt << std::endl;

        auto result = pipeline.processPrompt(prompt);
        REQUIRE(result.has_value());
        REQUIRE(result->operations.size() > 0);
        REQUIRE(result->daw_commands.size() > 0);

        std::cout << "✓ Complex operation passed with " << result->operations.size()
                  << " operations and " << result->daw_commands.size() << " DAW commands" << std::endl;
    }
}

TEST_CASE("Multilingual support integration", "[integration][multilingual]") {
    const char* api_key = std::getenv("OPENAI_API_KEY");
    if (!api_key) {
        SKIP("OPENAI_API_KEY not set - skipping integration tests");
    }

    magda::MAGDAPipeline pipeline(api_key);

    std::vector<std::pair<std::string, magda::OperationType>> multilingual_prompts = {
        // Spanish
        {"Crea una nueva pista llamada 'Guitarra Española'", magda::OperationType::CREATE_TRACK},
        {"Ajusta el volumen de 'Piano' a -3dB", magda::OperationType::SET_VOLUME},
        {"Añade reverb a la pista 'Bajo'", magda::OperationType::ADD_EFFECT},
        // French
        {"Ajoute une nouvelle piste audio nommée 'Batterie'", magda::OperationType::CREATE_TRACK},
        {"Réduis le volume de 'Synth' de 2dB", magda::OperationType::SET_VOLUME},
        {"Ajoute un filtre passe-bas à la piste 'Guitare'", magda::OperationType::ADD_EFFECT},
        // German
        {"Füge eine neue MIDI-Spur namens 'Klavier' hinzu", magda::OperationType::CREATE_TRACK},
        {"Erhöhe die Lautstärke der 'Bass'-Spur um 3dB", magda::OperationType::SET_VOLUME},
        {"Füge einen Chorus-Effekt zur 'Gitarre'-Spur hinzu", magda::OperationType::ADD_EFFECT},
        // Italian
        {"Crea una nuova traccia chiamata 'Chitarra Italiana'", magda::OperationType::CREATE_TRACK},
        {"Imposta il volume della traccia 'Piano' a -6dB", magda::OperationType::SET_VOLUME},
        {"Aggiungi riverbero alla traccia 'Basso'", magda::OperationType::ADD_EFFECT},
        // Portuguese
        {"Crie uma nova faixa chamada 'Guitarra Portuguesa'", magda::OperationType::CREATE_TRACK},
        {"Ajuste o volume da faixa 'Piano' para -3dB", magda::OperationType::SET_VOLUME},
        {"Adicione reverb à faixa 'Baixo'", magda::OperationType::ADD_EFFECT}
    };

    for (const auto& [prompt, expected_type] : multilingual_prompts) {
        std::cout << "\nTesting multilingual: " << prompt << std::endl;

        auto result = pipeline.processPrompt(prompt);
        REQUIRE(result.has_value());
        REQUIRE(result->operations.size() > 0);

        // Check for expected operation type
        bool has_expected_op = false;
        for (const auto& op : result->operations) {
            if (op.operation_type == expected_type) {
                has_expected_op = true;
                break;
            }
        }
        REQUIRE(has_expected_op);

        std::cout << "✓ Multilingual operation passed" << std::endl;
    }
}

TEST_CASE("Error handling integration", "[integration][error]") {
    const char* api_key = std::getenv("OPENAI_API_KEY");
    if (!api_key) {
        SKIP("OPENAI_API_KEY not set - skipping integration tests");
    }

    magda::MAGDAPipeline pipeline(api_key);

    std::vector<std::string> invalid_prompts = {
        "This is not a valid DAW command",
        "Random text that doesn't make sense",
        "123456789",
        ""
    };

    for (const auto& prompt : invalid_prompts) {
        std::cout << "\nTesting error handling: " << prompt << std::endl;

        auto result = pipeline.processPrompt(prompt);

        // Should either return no operations or handle gracefully
        if (result.has_value()) {
            if (result->operations.empty()) {
                std::cout << "✓ No operations identified (expected)" << std::endl;
            } else {
                std::cout << "Unexpected result with operations: " << result->operations.size() << std::endl;
            }
        } else {
            std::cout << "✓ Error handled gracefully" << std::endl;
        }
    }
}

File: python/tests/test_integration.py
"""
Integration tests for MAGDA that make real API calls.
Tests the full pipeline with various prompts including multilingual support.
"""

import os

import pytest

from magda.pipeline import MAGDAPipeline


class TestIntegration:
    """Integration tests with real API calls."""

    @pytest.fixture
    def pipeline(self):
        """Create a pipeline instance for testing."""
        api_key = os.getenv("OPENAI_API_KEY")
        if not api_key:
            pytest.skip("OPENAI_API_KEY not set")
        return MAGDAPipeline()

    # Individual track operation tests
    def test_create_track_basic(self, pipeline):
        """Test basic track creation."""
        prompt = "Create a new track called 'Guitar'"
        result = pipeline.process_prompt(prompt)

        assert result is not None, f"Pipeline failed for: {prompt}"
        assert "operations" in result, f"No operations found for: {prompt}"
        assert result["operations"], f"No operations identified for: {prompt}"

        operation_types = [op.get("type", "") for op in result["operations"]]
        assert "track" in operation_types, f"Expected track operation for: {prompt}"

    def test_create_audio_track(self, pipeline):
        """Test audio track creation."""
        prompt = "Add a new audio track named 'Bass'"
        result = pipeline.process_prompt(prompt)

        assert result is not None, f"Pipeline failed for: {prompt}"
        assert "operations" in result, f"No operations found for: {prompt}"
        assert result["operations"], f"No operations identified for: {prompt}"

        operation_types = [op.get("type", "") for op in result["operations"]]
        assert "track" in operation_types, f"Expected track operation for: {prompt}"

    def test_delete_track(self, pipeline):
        """Test track deletion."""
        prompt = "Delete the track 'Drums'"
        result = pipeline.process_prompt(prompt)

        assert result is not None, f"Pipeline failed for: {prompt}"
        assert "operations" in result, f"No operations found for: {prompt}"
        assert result["operations"], f"No operations identified for: {prompt}"

        operation_types = [op.get("type", "") for op in result["operations"]]
        assert "track" in operation_types, f"Expected track operation for: {prompt}"

    # Individual volume operation tests
    def test_set_volume(self, pipeline):
        """Test volume setting."""
        prompt = "Set the volume of track 'Guitar' to -6dB"
        result = pipeline.process_prompt(prompt)

        assert result is not None, f"Pipeline failed for: {prompt}"
        assert "operations" in result, f"No operations found for: {prompt}"
        assert result["operations"], f"No operations identified for: {prompt}"

        operation_types = [op.get("type", "") for op in result["operations"]]
        assert "volume" in operation_types, f"Expected volume operation for: {prompt}"

    def test_increase_volume(self, pipeline):
        """Test volume increase."""
        prompt = "Increase the volume of 'Bass' by 3dB"
        result = pipeline.process_prompt(prompt)

        assert result is not None, f"Pipeline failed for: {prompt}"
        assert "operations" in result, f"No operations found for: {prompt}"
        assert result["operations"], f"No operations identified for: {prompt}"

        operation_types = [op.get("type", "") for op in result["operations"]]
        assert "volume" in operation_types, f"Expected volume operation for: {prompt}"

    def test_mute_track(self, pipeline):
        """Test track muting."""
        prompt = "Mute the 'Drums' track"
        result = pipeline.process_prompt(prompt)

        assert result is not None, f"Pipeline failed for: {prompt}"
        assert "operations" in result, f"No operations found for: {prompt}"
        assert result["operations"], f"No operations identified for: {prompt}"

        operation_types = [op.get("type", "") for op in result["operations"]]
        # Could be either track or volume operation
        assert any(op_type in operation_types for op_type in ["track", "volume"]), (
            f"Expected track or volume operation for: {prompt}"
        )

    # Individual effect operation tests
    def test_add_reverb(self, pipeline):
        """Test adding reverb effect."""
        prompt = "Add reverb to the 'Guitar' track"
        result = pipeline.process_prompt(prompt)

        assert result is not None, f"Pipeline failed for: {prompt}"
        assert "operations" in result, f"No operations found for: {prompt}"
        assert result["operations"], f"No operations identified for: {prompt}"

        operation_types = [op.get("type", "") for op in result["operations"]]
        assert "effect" in operation_types, f"Expected effect operation for: {prompt}"

    def test_remove_delay(self, pipeline):
        """Test removing delay effect."""
        prompt = "Remove the delay from 'Bass'"
        result = pipeline.process_prompt(prompt)

        assert result is not None, f"Pipeline failed for: {prompt}"
        assert "operations" in result, f"No operations found for: {prompt}"
        assert result["operations"], f"No operations identified for: {prompt}"

        operation_types = [op.get("type", "") for op in result["operations"]]
        assert "effect" in operation_types, f"Expected effect operation for: {prompt}"

    def test_set_reverb_wet_level(self, pipeline):
        """Test setting reverb wet level."""
        prompt = "Set the reverb wet level to 50% on 'Guitar'"
        result = pipeline.process_prompt(prompt)

        assert result is not None, f"Pipeline failed for: {prompt}"
        assert "operations" in result, f"No operations found for: {prompt}"
        assert result["operations"], f"No operations identified for: {prompt}"

        operation_types = [op.get("type", "") for op in result["operations"]]
        # Could be effect parameter or volume
        assert any(op_type in operation_types for op_type in ["effect", "volume"]), (
            f"Expected effect or volume operation for: {prompt}"
        )

    # Individual clip operation tests
    def test_create_clip(self, pipeline):
        """Test clip creation."""
        prompt = "Create a 4-bar clip on track 'Guitar'"
        result = pipeline.process_prompt(prompt)

        assert result is not None, f"Pipeline failed for: {prompt}"
        assert "operations" in result, f"No operations found for: {prompt}"
        assert result["operations"], f"No operations identified for: {prompt}"

        operation_types = [op.get("type", "") for op in result["operations"]]
        assert "clip" in operation_types, f"Expected clip operation for: {prompt}"

    def test_delete_clip(self, pipeline):
        """Test clip deletion."""
        prompt = "Delete the clip at bar 8 on 'Bass'"
        result = pipeline.process_prompt(prompt)

        assert result is not None, f"Pipeline failed for: {prompt}"
        assert "operations" in result, f"No operations found for: {prompt}"
        assert result["operations"], f"No operations identified for: {prompt}"

        operation_types = [op.get("type", "") for op in result["operations"]]
        assert "clip" in operation_types, f"Expected clip operation for: {prompt}"

    def test_move_clip(self, pipeline):
        """Test clip movement."""
        prompt = "Move the clip from bar 4 to bar 12 on 'Drums'"
        result = pipeline.process_prompt(prompt)

        assert result is not None, f"Pipeline failed for: {prompt}"
        assert "operations" in result, f"No operations found for: {prompt}"
        assert result["operations"], f"No operations identified for: {prompt}"

        operation_types = [op.get("type", "") for op in result["operations"]]
        assert "clip" in operation_types, f"Expected clip operation for: {prompt}"

    # Individual MIDI operation tests
    def test_add_midi_chord(self, pipeline):
        """Test adding MIDI chord."""
        prompt = "Add a C major chord at bar 1 on 'Piano'"
        result = pipeline.process_prompt(prompt)

        assert result is not None, f"Pipeline failed for: {prompt}"
        assert "operations" in result, f"No operations found for: {prompt}"
        assert result["operations"], f"No operations identified for: {prompt}"

        operation_types = [op.get("type", "") for op in result["operations"]]
        assert "midi" in operation_types, f"Expected MIDI operation for: {prompt}"

    def test_delete_midi_note(self, pipeline):
        """Test deleting MIDI note."""
        prompt = "Delete the note at beat 2.5 on 'Bass'"
        result = pipeline.process_prompt(prompt)

        assert result is not None, f"Pipeline failed for: {prompt}"
        assert "operations" in result, f"No operations found for: {prompt}"
        assert result["operations"], f"No operations identified for: {prompt}"

        operation_types = [op.get("type", "") for op in result["operations"]]
        assert "midi" in operation_types, f"Expected MIDI operation for: {prompt}"

    def test_change_midi_velocity(self, pipeline):
        """Test changing MIDI velocity."""
        prompt = "Change the velocity of the note at bar 4 to 80"
        result = pipeline.process_prompt(prompt)

        assert result is not None, f"Pipeline failed for: {prompt}"
        assert "operations" in result, f"No operations found for: {prompt}"
        assert result["operations"], f"No operations identified for: {prompt}"

        operation_types = [op.get("type", "") for op in result["operations"]]
        assert "midi" in operation_types, f"Expected MIDI operation for: {prompt}"

    # Multilingual tests using parametrize
    @pytest.mark.parametrize(
        "prompt,expected_type",
        [
            ("Crea una nueva pista llamada 'Guitarra'", "track"),
            ("Ajusta el volumen de la pista 'Bajo' a -3dB", "volume"),
            ("Ajoute une nouvelle piste audio nommée 'Batterie'", "track"),
            ("Füge einen Reverb-Effekt zur 'Gitarre'-Spur hinzu", "effect"),
            ("トラック「ベース」の音量を-6dBに設定してください", "volume"),
        ],
    )
    def test_multilingual_operations(self, pipeline, prompt, expected_type):
        """Test multilingual operation identification."""
        result = pipeline.process_prompt(prompt)

        assert result is not None, f"Multilingual processing failed for: {prompt}"
        assert "operations" in result, f"No operations found for: {prompt}"
        assert result["operations"], f"No operations identified for: {prompt}"

        operation_types = [op.get("type", "") for op in result["operations"]]
        assert expected_type in operation_types, (
            f"Expected {expected_type} in {operation_types} for: {prompt}"
        )

    # Complex multi-operation tests
    def test_multi_operation_track_setup(self, pipeline):
        """Test complex multi-operation prompt."""
        prompt = (
            "Create a new track called 'Lead', set its volume to -3dB, and add reverb"
        )
        result = pipeline.process_prompt(prompt)

        assert result is not None, f"Pipeline failed for: {prompt}"
        assert "operations" in result, f"No operations found for: {prompt}"
        assert result["operations"], f"No operations identified for: {prompt}"

        # Should identify multiple operations
        operation_types = [op.get("type", "") for op in result["operations"]]
        assert "track" in operation_types, f"Expected track operation for: {prompt}"

    def test_midi_track_complex_setup(self, pipeline):
        """Test complex MIDI track setup."""
        prompt = (
            "Add a new MIDI track 'Piano', create a 4-bar clip, and add a C major chord"
        )
        result = pipeline.process_prompt(prompt)

        assert result is not None, f"Pipeline failed for: {prompt}"
        assert "operations" in result, f"No operations found for: {prompt}"
        assert result["operations"], f"No operations identified for: {prompt}"

        # Should identify multiple operations
        operation_types = [op.get("type", "") for op in result["operations"]]
        assert "track" in operation_types, f"Expected track operation for: {prompt}"

    # Error handling tests
    @pytest.mark.parametrize(
        "invalid_prompt",
        [
            "This is not a valid DAW command",
            "Random text that doesn't make sense",
            "123456789",
            "",
        ],
    )
    def test_error_handling_invalid_prompts(self, pipeline, invalid_prompt):
        """Test error handling with invalid prompts - should handle gracefully."""
        result = pipeline.process_prompt(invalid_prompt)
        # Should either return None or have no operations
        if result is not None:
            if "operations" in result and not result["operations"]:
                pass  # Expected: no operations identified
            else:
                # If we get unexpected operations, that's fine too - just log it
                print(f"Unexpected result: {result}")

    def test_error_handling_timeout(self, pipeline):
        """Test error handling with timeout scenarios."""
        # This test would require mocking the timeout behavior
        # For now, we'll test that the pipeline doesn't crash on very long prompts
        long_prompt = "Create a track called 'Test' " * 1000  # Very long prompt
        result = pipeline.process_prompt(long_prompt)
        # Should handle gracefully without crashing
        assert result is not None or result is None  # Either is acceptable

    # Full pipeline integration tests
    @pytest.mark.parametrize(
        "complex_prompt",
        [
            "Create a new track called 'Lead Guitar', set its volume to -3dB, and add reverb with 40% wet level",
            "Add a new MIDI track 'Piano', create a 4-bar clip, and add a C major chord at bar 1",
            "Crea una pista llamada 'Bajo', ajusta el volumen a -6dB, y añade un efecto de compresión",
            "Ajoute une piste audio 'Batterie', crée un clip de 8 mesures, et ajoute un filtre passe-haut",
        ],
    )
    def test_full_pipeline_complex(self, pipeline, complex_prompt):
        """Test the complete pipeline end-to-end with complex prompts."""
        result = pipeline.process_prompt(complex_prompt)

        assert result is not None, f"Pipeline failed for: {complex_prompt}"
        assert "operations" in result, f"No operations found for: {complex_prompt}"
        assert "daw_commands" in result, (
            f"No DAW commands generated for: {complex_prompt}"
        )
        assert result["operations"], f"No operations identified for: {complex_prompt}"


if __name__ == "__main__":
    # Run integration tests directly
    pytest.main([__file__, "-v", "-s"])
